### PR TITLE
PaywallData validation

### PR DESCRIPTION
**Pull request by vegaro**
Add validation to paywall data and creates default template if validation fails

### CODE REVIEW

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 1]
```diff
@@ -8,23 +8,21 @@ import com.revenuecat.purchases.ui.revenuecatui.data.processed.VariableDataProvi
...
-            locale = Locale.getDefault(), // TODO-PAYWALLS: Use the correct locale
-        )
-        PaywallViewState.Loaded(
-            templateConfiguration = templateConfiguration,
-            selectedPackage = templateConfiguration.packages.default,
-        )
-    } catch (e: Exception) {
-        PaywallViewState.Error("Error creating paywall: ${e.message}")
-    }
+    val templateConfiguration = TemplateConfigurationFactory.create(
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:16]: Looks like this method is still calling `PaywallTemplate.fromId(paywallData.templateName)` and failing if not found.
To avoid that I made the validation result produce the processed `PaywallTemplate`, so then we could inject it here. What do you think?

Comment (NachoSoto) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:16]: Note that in iOS the `ValidationResult` also has the `displayablePaywall: PaywallData` which is how we can reuse the configured `packages` even if there are validation errors.

**Comment by NachoSoto** [comment]
This is looking good 👍🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.revenuecat.purchases.ui.revenuecatui.utils
+
+internal sealed class Result<out A, out B> {
```
Comment (tonidero) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt:3]: I noticed it later, but kotlin has its own `Result` class that we can also use...Should we use that?

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/InternalPaywallView.kt [hunk 3]
```diff
@@ -63,6 +64,7 @@ private fun getPaywallViewModel(offering: Offering?, listener: PaywallViewListen
             LocalContext.current.applicationContext.toAndroidContext(),
             offering,
             listener,
+            MaterialTheme.colors,
```
Comment (tonidero) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/InternalPaywallView.kt:67]: We should check if light/dark theme colors get automatically updated in the default paywall. My strong bet is that no. In that case, we should maybe store the last state in the view model so we can update the state when it changes, but we can do that in another PR.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt [hunk 4]
```diff
@@ -48,6 +50,8 @@ internal class PaywallViewModelImpl(
     applicationContext: ApplicationContext,
     private val offering: Offering?,
     private val listener: PaywallViewListener?,
+    private val packageName: String,
```
Comment (tonidero) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt:53]: An alternative to passing the package name here would be to add a method in the `ApplicationContext` to get it... I'm ok keeping it as a separate parameter though if you prefer it.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 5]
```diff
@@ -1,28 +1,116 @@
...
         )
-        PaywallViewState.Loaded(
-            templateConfiguration = templateConfiguration,
-            selectedPackage = templateConfiguration.packages.default,
+        is Result.Success -> ValidationResult(
+            paywallData,
+            result.value,
+            null,
         )
-    } catch (e: Exception) {
```
Comment (tonidero) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:25]: Since we're not doing this anymore, it may crash in the `TemplateConfigurationFactory.create` since we `require` a that there is at least some packages. Should we perform that validation before to avoid the crash?

Comment (vegaro) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:28]: Right, I missed those `require`

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt (hunk 2 above)
Comment (vegaro) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt:3]: Oh I thought you couldn't use it as a return type but it looks like since 1.5 you can! https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/result.md#summary

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/InternalPaywallView.kt [hunk 6]
```diff
@@ -63,12 +64,15 @@ private fun getPaywallViewModel(
...
 ): PaywallViewModel {
+    val applicationContext = LocalContext.current.applicationContext
     return viewModel<PaywallViewModelImpl>(
         factory = PaywallViewModelFactory(
-            LocalContext.current.applicationContext.toAndroidContext(),
+            applicationContext.toAndroidContext(),
             mode,
             offering,
             listener,
+            applicationContext.packageName,
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/InternalPaywallView.kt:74]: Yeah we're already passing `ApplicationContext` here, so let's add a method to that interface instead?

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt (hunk 4 above)
Comment (NachoSoto) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt:55]: I agree 👍🏻 

Also I know this matches the Android property, but could we change this so it doesn't get confused with `RevenueCat.Package`? It's kinda confusing now hehe. Maybe `applicationManifestName`?

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/TemplateConfigurationFactory.kt [hunk 7]
```diff
@@ -10,37 +10,35 @@ internal object TemplateConfigurationFactory {
     fun create(
         variableDataProvider: VariableDataProvider,
         mode: PaywallViewMode,
-        paywallData: PaywallData,
+        validatedPaywallData: PaywallData,
         packages: List<Package>,
         activelySubscribedProductIdentifiers: Set<String>,
+        template: PaywallTemplate,
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/TemplateConfigurationFactory.kt:16]: 👍🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/VariableProcessor.kt [hunk 8]
```diff
@@ -7,18 +7,38 @@ import java.util.Locale
...
         variableDataProvider: VariableDataProvider,
         originalString: String,
         rcPackage: Package,
         locale: Locale,
     ): String {
-        var resultString = originalString
-        REGEX.findAll(originalString).toList().reversed().forEach { matchResult ->
+        val resultString = handleVariablesAndReplace(originalString) { variable ->
+            val replacement = variableValue(variableDataProvider, variable, rcPackage, locale)
+            replacement
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/VariableProcessor.kt:27]: Maybe
```suggestion
            variableValue(variableDataProvider, variable, rcPackage, locale)
```

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/errors/PaywallValidationError.kt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package com.revenuecat.purchases.ui.revenuecatui.errors
+
+sealed class PaywallValidationError {
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/errors/PaywallValidationError.kt:3]: 😻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 10]
```diff
@@ -1,31 +1,117 @@
...
+}
+
+private fun PaywallData.validate(): Result<PaywallTemplate, PaywallValidationError> {
+    val (_, localizedConfiguration) = localizedConfiguration
+
+    localizedConfiguration.validate().takeIf { it != null }?.let { Result.Error(it) }
+
+    val template = validateTemplate() ?: return Result.Error(PaywallValidationError.InvalidTemplate(templateName))
+
+    // TODO-PAYWALLS: Validate icons
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:64]: I added them in #1274, but you can add this in a separate PR 👍🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt (hunk 1 above)
Comment (NachoSoto) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:77]: Addressed 👍🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 11]
```diff
@@ -1,31 +1,117 @@
...
+    return Result.Success(template)
+}
 
 @Suppress("ReturnCount", "TooGenericExceptionCaught")
 internal fun Offering.toPaywallViewState(
     variableDataProvider: VariableDataProvider,
     mode: PaywallViewMode,
+    validatedPaywallData: PaywallData,
+    template: PaywallTemplate,
+    @Suppress("UNUSED_PARAMETER") error: PaywallValidationError?,
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:75]: We have a linear for this: https://linear.app/revenuecat/issue/PWL-267/display-error-when-displaying-default-paywall-on-debug-builds 👍🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 12]
```diff
@@ -8,23 +8,21 @@ import com.revenuecat.purchases.ui.revenuecatui.data.processed.VariableDataProvi
...
-    }
+    val templateConfiguration = TemplateConfigurationFactory.create(
+        variableDataProvider = variableDataProvider,
+        mode = PaywallViewMode.FULL_SCREEN,
+        paywallData = paywallData,
+        packages = availablePackages,
+        activelySubscribedProductIdentifiers = emptySet(), // TODO-PAYWALLS: Check for active subscriptions
+        locale = Locale.getDefault(), // TODO-PAYWALLS: Use the correct locale
+    )
+    return PaywallViewState.Loaded(
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:85]: Maybe we can change the return of this function to `PaywallViewState.Loaded` now?

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 13]
```diff
@@ -1,31 +1,117 @@
...
+        template,
+    )
+    return PaywallViewState.Loaded(
+        templateConfiguration = templateConfiguration,
+        selectedPackage = templateConfiguration.packages.default,
+    )
+    // TODO-PAYWALLS: Handle error
+}
+
+private fun PaywallData.LocalizedConfiguration.validate(): PaywallValidationError.InvalidVariables? {
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:92]: Maybe
```suggestion
private fun PaywallData.LocalizedConfiguration.validateVariables(): PaywallValidationError.InvalidVariables? {
```

**Review by NachoSoto — COMMENTED** [review]
I think this is pretty much ready! 👏🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt (hunk 5 above)
Comment (vegaro) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:28]: [`ec1a4f4` (#1273)](https://github.com/RevenueCat/purchases-android/pull/1273/commits/ec1a4f4627cd38195ec4d60ef925d5e768aa2968) and [`6bb5ebc` (#1273)](https://github.com/RevenueCat/purchases-android/pull/1273/commits/6bb5ebcfee6132a6cd4fea19abd12ec30567d482)

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/InternalPaywallView.kt (hunk 6 above)
Comment (vegaro) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/InternalPaywallView.kt:74]: Yes, let's do that, much cleaner

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt (hunk 11 above)
Comment (vegaro) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:61]: ok, I will remove the parameter and change the return type

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt (hunk 4 above)
Comment (vegaro) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt]: I named it `applicationPackageName`, I think it's less confusing. I find `applicationManifestName` strange 🤷

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt [hunk 14]
```diff
@@ -143,7 +149,13 @@ internal class PaywallViewModelImpl(
         }
     }
 
+    @Suppress("TooGenericExceptionCaught")
     private fun Offering.calculateState(): PaywallViewState {
-        return toPaywallViewState(variableDataProvider, mode)
+        if (availablePackages.isEmpty()) {
+            return PaywallViewState.Error("No packages available")
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt:155]: `RevenueCat` ensures that `Offering` never has empty packages, but this doesn't hurt here 👍🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/TestData.kt [hunk 15]
```diff
@@ -249,16 +250,24 @@ internal class MockApplicationContext : ApplicationContext {
...
     override val state: StateFlow<PaywallViewState>
         get() = _state.asStateFlow()
-    private val _state =
-        MutableStateFlow(offering.toPaywallViewState(VariableDataProvider(MockApplicationContext()), mode))
+    private val _state = MutableStateFlow(
+        offering.toPaywallViewState(
+            VariableDataProvider(MockApplicationContext()),
+            mode,
+            TestData.template2,
+            PaywallTemplate.TEMPLATE_2,
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/TestData.kt:268]: This is no longer getting the paywall from `offering`. Since this is just for test data, maybe do:
```suggestion
            offering.paywall!!,
            PaywallTemplate.fromId(offering.paywall!!)!!,
```

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 16]
```diff
@@ -1,31 +1,101 @@
...
+import com.revenuecat.purchases.ui.revenuecatui.extensions.defaultTemplate
+import com.revenuecat.purchases.ui.revenuecatui.utils.Result
+
+internal fun Offering.validatedPaywall(
+    packageName: String,
+    currentColors: Colors,
+): PaywallValidationResult {
+    val paywallData = this.paywall
+        ?: return PaywallValidationResult(
+            PaywallData.createDefault(packages = availablePackages, packageName, currentColors),
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:23]: 👏🏻

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 17]
```diff
@@ -1,31 +1,101 @@
...
+            PaywallData.defaultTemplate,
+            PaywallValidationError.MissingPaywall,
+        )
+
+    return when (val result = paywallData.validate()) {
+        is Result.Error -> PaywallValidationResult(
+            PaywallData.createDefault(packages = availablePackages, packageName, currentColors),
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:30]: This is the important part: instead of `availablePackages`, this should use `paywallData.packages`. So even though the paywall wasn't valid, we can still show only the packages that the developer wanted to display.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 18]
```diff
@@ -1,31 +1,101 @@
...
+        template,
+    )
+    return PaywallViewState.Loaded(
+        templateConfiguration = templateConfiguration,
+        selectedPackage = templateConfiguration.packages.default,
+    )
+}
+
+private fun PaywallData.LocalizedConfiguration.validateVariables(): PaywallValidationError.InvalidVariables? {
+    fun String?.validateVariablesInProperty(): Set<String> {
```
Comment (NachoSoto) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:77]: Awesome

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt (hunk 2 above)
Comment (NachoSoto) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt:3]: So we can remove this now?

**Review by NachoSoto — COMMENTED** [review]
Just a few minor things left, but this is perfect 🎉

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt (hunk 2 above)
Comment (vegaro) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/utils/Result.kt]: Yes, I just replaced it with Kotlin's Result 😄

**Comment by vegaro** [comment]
Fixed the light/dark colors. We will default to the current color scheme now, when it changes, and we react to the change

**Comment by vegaro** [comment]
@tonidero I will do the icons validation in another PR so we can merge this one. Do you mind taking a look?

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/TemplateConfigurationFactory.kt [hunk 19]
```diff
@@ -10,37 +10,35 @@ internal object TemplateConfigurationFactory {
     fun create(
         variableDataProvider: VariableDataProvider,
         mode: PaywallViewMode,
-        paywallData: PaywallData,
+        validatedPaywallData: PaywallData,
```
Comment (tonidero) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/TemplateConfigurationFactory.kt:13]: I'm wondering if this should keep being `paywallData` to not be confused with `ValidatedPaywallResult`... I think it's ok though.

**Review by tonidero — APPROVED** [review]
Looks great!
